# Data Collecting News API

### News API Data collection JSON

In [127]:
import requests
import json
from datetime import datetime, timedelta

def fetch_news(api_key, query, from_date, to_date, page_size=100):
    url = f'https://newsapi.org/v2/everything?q={query}&from={from_date}&to={to_date}&pageSize={page_size}&apiKey={api_key}'
    response = requests.get(url)
    data = response.json()
    return data

def save_to_json(data, filename):
    with open(filename, 'w') as f:
        json.dump(data, f, indent=4)

if __name__ == "__main__":
    API_KEY = 'YOUR API KEY'
    QUERY = 'Apple'
    
    # Define the date range for the last month
    to_date = datetime.now().strftime('%Y-%m-%d')
    from_date = (datetime.now() - timedelta(days=30)).strftime('%Y-%m-%d')
    
    news_data = fetch_news(API_KEY, QUERY, from_date, to_date)
    
    # Generate a filename with the current date
    date_str = datetime.now().strftime('%Y-%m-%d')
    filename = f'{QUERY}{date_str}.json'
    
    #save_to_json(news_data, filename)
    print(f'News data saved to {filename}')


News data saved to Apple2024-09-11.json


In [128]:
import requests
import pandas as pd
from datetime import datetime, timedelta

def fetch_news(api_key, query, from_date, to_date, page_size=100):
    url = f'https://newsapi.org/v2/everything?q={query}&from={from_date}&to={to_date}&pageSize={page_size}&apiKey={api_key}'
    response = requests.get(url)
    data = response.json()
    return data

def save_to_csv(data, filename):
    df = pd.DataFrame(data['articles'])
    df.to_csv(filename, index=False)

if __name__ == "__main__":
    API_KEY = '5b654d218f124beeab121df62f4c4371'
    QUERY = 'Microsoft'
    
    # Define the date range for the last month
    to_date = datetime.now().strftime('%Y-%m-%d')
    from_date = (datetime.now() - timedelta(days=30)).strftime('%Y-%m-%d')
    
    news_data = fetch_news(API_KEY, QUERY, from_date, to_date)
    
    # Generate a filename with the current date
    date_str = datetime.now().strftime('%Y-%m-%d')
    filename = f'{QUERY}{date_str}.csv'
    
    data = news_data


In [129]:
import requests
import pandas as pd
from datetime import datetime, timedelta

def fetch_news(api_key, query, from_date, to_date, page_size=100):
    url = f'https://newsapi.org/v2/everything?q={query}&from={from_date}&to={to_date}&pageSize={page_size}&apiKey={api_key}'
    response = requests.get(url)
    data = response.json()
    return data

def parse_articles(data):
    articles = data.get('articles', [])
    
    # Extract relevant fields and create a structured DataFrame
    parsed_data = []
    for article in articles:
        parsed_data.append({
            'source': article['source']['name'],
            'author': article.get('author'),
            'title': article.get('title'),
            'description': article.get('description'),
            'url': article.get('url'),
            'publishedAt': article.get('publishedAt'),
            'content': article.get('content')
        })
    
    return pd.DataFrame(parsed_data)

if __name__ == "__main__":
    API_KEY = '5b654d218f124beeab121df62f4c4371'
    QUERY = 'Microsoft'
    
    # Define the date range for the last month
    to_date = datetime.now().strftime('%Y-%m-%d')
    from_date = (datetime.now() - timedelta(days=30)).strftime('%Y-%m-%d')
    
    # Fetch news data
    news_data = fetch_news(API_KEY, QUERY, from_date, to_date)
    
    # Convert data into a DataFrame and assign it to the variable 'data'
    data = parse_articles(news_data)
    
    # Display the DataFrame
    data


## Data cleaning

* Remove duplicates
* Deal with missing value
* In a correct format
* Correct error type of out range value
* Address outliers
* Validate the data ensure it accurate and reliable

In [130]:
import pandas as pd



# Print basic information about the dataset
print("Basic Information:")
print(data.info())



Basic Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   source       100 non-null    object
 1   author       95 non-null     object
 2   title        100 non-null    object
 3   description  99 non-null     object
 4   url          100 non-null    object
 5   publishedAt  100 non-null    object
 6   content      100 non-null    object
dtypes: object(7)
memory usage: 5.6+ KB
None


### Checking summary statistics

In [131]:
# Print summary statistics
print("\nSummary Statistics:")
print(data.describe(include='all'))





Summary Statistics:
             source        author      title description                  url  \
count           100            95        100          99                  100   
unique           21            63         97          96                   97   
top     Gizmodo.com  Luis Miranda  [Removed]   [Removed]  https://removed.com   
freq             13             8          4           4                    4   

                 publishedAt  \
count                    100   
unique                    97   
top     1970-01-01T00:00:00Z   
freq                       4   

                                                  content  
count                                                 100  
unique                                                 93  
top     If you click 'Accept all', we and our partners...  
freq                                                    5  


## Missing values

In [132]:
# Check for missing values
print("\nMissing Values:")
print(data.isnull().sum())




Missing Values:
source         0
author         5
title          0
description    1
url            0
publishedAt    0
content        0
dtype: int64


### Display rows with missing values and deal with it

In [133]:
rows_with_missing_values = data[data.isnull().any(axis=1)]

# Print the rows with missing values
print("Rows with Missing Values:")
rows_with_missing_values

Rows with Missing Values:


source author  \
6             [Removed]   None   
7             [Removed]   None   
15            [Removed]   None   
30            [Removed]   None   
54  Yahoo Entertainment   None   

                                                title description  \
6                                           [Removed]   [Removed]   
7                                           [Removed]   [Removed]   
15                                          [Removed]   [Removed]   
30                                          [Removed]   [Removed]   
54  Microsoft to host cybersecurity summit after C...        None   

                                                  url           publishedAt  \
6                                 https://removed.com  1970-01-01T00:00:00Z   
7                                 https://removed.com  1970-01-01T00:00:00Z   
15                                https://removed.com  1970-01-01T00:00:00Z   
30                                https://removed.com  1970-01-01T00:00:00Z   
54  https://consent.yahoo.com/v2/collectConsent?se...  2024-08-23T15:28:45Z   

                                              content  
6                                           [Removed]  
7                                           [Removed]  
15                                          [Removed]  
30                                          [Removed]  
54  If you click 'Accept all', we and our partners...

So far the decision is to just keep the missing values because as it is but if the whole rows has missing values then we will remove it 

In [134]:
# Fill missing values with placeholders
data['author'].fillna('Unknown', inplace=True)
data['description'].fillna('No description available', inplace=True)
data['content'].fillna('No content available', inplace=True)

# Remove rows where the title is '[Removed]'
data = data[data['title'] != '[Removed]']

# Check for missing values
print("\nMissing Values:")
print(data.isnull().sum())


# Print the cleaned data
print(data)


Missing Values:
source         0
author         0
title          0
description    0
url            0
publishedAt    0
content        0
dtype: int64
                 source            author  \
0                 Wired  Eric Ravenscraft   
1             The Verge        Tom Warren   
2             The Verge        Tom Warren   
3   Yahoo Entertainment        Steve Dent   
4   Yahoo Entertainment        Jeremy Gan   
..                  ...               ...   
95                Wired       Steven Levy   
96           Xataka.com   Ricardo Aguilar   
97       Digital Trends    Fionna Agomuoh   
98       Digital Trends   Monica J. White   
99       Digital Trends        Judy Sanhz   

                                                title  \
0   Only Microsoft Can Save Windows-Powered Handhe...   
1   Microsoft confirms its Paint 3D app is being d...   
2   Microsoft’s new Qualcomm-powered Surface devic...   
3   PUBG: Battlegrounds publisher acquires Hi-Fi R...   
4   Microsoft is testing 

C:\Users\muham\AppData\Local\Temp\ipykernel_12952\1991064513.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['author'].fillna('Unknown', inplace=True)
C:\Users\muham\AppData\Local\Temp\ipykernel_12952\1991064513.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For examp

### Checking duplicate rows

In [135]:
# Check for duplicate rows
print("\nDuplicate Rows:")
print(data.duplicated().sum())




Duplicate Rows:
0


### Change the data format so that it is more readable

In [136]:
# Check for any unusual data types or inconsistencies
print("\nData Types:")
print(data.dtypes)




Data Types:
source         object
author         object
title          object
description    object
url            object
publishedAt    object
content        object
dtype: object


In [137]:
# Fix date-time format
data['publishedAt'] = pd.to_datetime(data['publishedAt']).dt.strftime('%Y-%m-%d %H:%M:%S')

# Simplify the 'source' column
data['source'] = data['source'].apply(lambda x: eval(x).get('name', 'Unknown') if isinstance(x, str) else 'Unknown')

# Print the cleaned data

NameError: name 'Wired' is not defined

In [138]:
data

source            author  \
0                 Wired  Eric Ravenscraft   
1             The Verge        Tom Warren   
2             The Verge        Tom Warren   
3   Yahoo Entertainment        Steve Dent   
4   Yahoo Entertainment        Jeremy Gan   
..                  ...               ...   
95                Wired       Steven Levy   
96           Xataka.com   Ricardo Aguilar   
97       Digital Trends    Fionna Agomuoh   
98       Digital Trends   Monica J. White   
99       Digital Trends        Judy Sanhz   

                                                title  \
0   Only Microsoft Can Save Windows-Powered Handhe...   
1   Microsoft confirms its Paint 3D app is being d...   
2   Microsoft’s new Qualcomm-powered Surface devic...   
3   PUBG: Battlegrounds publisher acquires Hi-Fi R...   
4   Microsoft is testing a new on-screen keyboard ...   
..                                                ...   
95      The NSA Has a Podcast—Here's How to Decode It   
96  Microsoft bloquea el truco que permitía actual...   
97  Microsoft OneDrive vs. Dropbox: Which is the b...   
98  It took Microsoft 30 years to change this Wind...   
99  Microsoft backs off on pressuring Windows 10 u...   

                                          description  \
0   The dream of running all your Steam, Epic, and...   
1   Microsoft is removing its Paint 3D app from it...   
2   Microsoft thinks Windows on Arm is ready for b...   
3   Under a cloud of studio closures and layoffs, ...   
4   Microsoft has been working on improving small-...   
..                                                ...   
95  The spy agency that dared not speak its name i...   
96  Microsoft quiere que los usuarios salten a Win...   
97  This comparison guide will tell you if Dropbox...   
98  After 30 years, Microsoft is finally changing ...   
99  Microsoft will do an about-face with putting a...   

                                                  url          publishedAt  \
0   https://www.wired.com/story/only-microsoft-can...  2024-08-31 13:00:00   
1   https://www.theverge.com/2024/8/12/24218450/mi...  2024-08-12 16:08:27   
2   https://www.theverge.com/2024/9/4/24235594/mic...  2024-09-04 11:00:00   
3   https://consent.yahoo.com/v2/collectConsent?se...  2024-08-12 12:00:05   
4   https://consent.yahoo.com/v2/collectConsent?se...  2024-09-04 13:50:58   
..                                                ...                  ...   
95  https://www.wired.com/story/the-nsa-has-a-podc...  2024-09-06 13:00:00   
96  https://www.xataka.com/aplicaciones/microsoft-...  2024-08-22 10:01:30   
97  https://www.digitaltrends.com/computing/micros...  2024-08-27 00:10:01   
98  https://www.digitaltrends.com/computing/micros...  2024-08-16 18:42:59   
99  https://www.digitaltrends.com/computing/one-ma...  2024-08-15 23:28:14   

                                              content  
0   Microsoft also requires some software to come ...  
1   Image: The Verge\r\n\n \n\n Paint 3D, once Mic...  
2   Photo by Chris Welch / The Verge\r\n\n \n\n Mi...  
3   If you click 'Accept all', we and our partners...  
4   If you click 'Accept all', we and our partners...  
..                                                ...  
95  My first story for WIREDyep, 31 years agolooke...  
96  Microsoft quiere que los usuarios salten a Win...  
97  Dropbox and OneDrive are two of the most popul...  
98  In 1996, Microsoft introduced the FAT32 file s...  
99  Microsoft has announced that it will ease up t...  

[96 rows x 7 columns]

In [139]:
# Convert 'publishedAt' to datetime format
data['publishedAt'] = pd.to_datetime(data['publishedAt'])

# Extract the date in 'yyyy-mm-dd' format
data['date'] = data['publishedAt'].dt.strftime('%d/%m/%Y')

print(data)

                 source            author  \
0                 Wired  Eric Ravenscraft   
1             The Verge        Tom Warren   
2             The Verge        Tom Warren   
3   Yahoo Entertainment        Steve Dent   
4   Yahoo Entertainment        Jeremy Gan   
..                  ...               ...   
95                Wired       Steven Levy   
96           Xataka.com   Ricardo Aguilar   
97       Digital Trends    Fionna Agomuoh   
98       Digital Trends   Monica J. White   
99       Digital Trends        Judy Sanhz   

                                                title  \
0   Only Microsoft Can Save Windows-Powered Handhe...   
1   Microsoft confirms its Paint 3D app is being d...   
2   Microsoft’s new Qualcomm-powered Surface devic...   
3   PUBG: Battlegrounds publisher acquires Hi-Fi R...   
4   Microsoft is testing a new on-screen keyboard ...   
..                                                ...   
95      The NSA Has a Podcast—Here's How to Decode It   
96 

In [140]:
# Ensure the 'date' column is of type datetime.date
data['date'] = pd.to_datetime(data['date'])

C:\Users\muham\AppData\Local\Temp\ipykernel_12952\1045107494.py:2: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  data['date'] = pd.to_datetime(data['date'])


In [141]:
print(data.dtypes)


source                 object
author                 object
title                  object
description            object
url                    object
publishedAt    datetime64[ns]
content                object
date           datetime64[ns]
dtype: object


In [142]:
clean_file = data

# Merge Financial Data and News Data

In [143]:
import yfinance as yf
import pandas as pd

# Define the stock ticker and the date range
ticker = 'MSFT'
start_date = '2024-01-01'
end_date = '2024-09-11'

# Fetch the stock data from Yahoo Finance
stock_df = yf.download(ticker, start=start_date, end=end_date)

# Reset the index to have 'Date' as a column
stock_df.reset_index(inplace=True)

# Ensure 'Date' is in datetime format
stock_df['Date'] = pd.to_datetime(stock_df['Date'])

# Drop the 'High', 'Low', and 'Volume' columns
stock_df.drop(columns=['High', 'Low', 'Volume'], inplace=True)

# Sort by date to ensure correct shifting
stock_df.sort_values(by='Date', inplace=True)

# Create columns for the next 2 stock prices
stock_df['Next_Stock_Price_1'] = stock_df['Close'].shift(-1)
stock_df['Next_Stock_Price_2'] = stock_df['Close'].shift(-2)

# Remove rows with NaN values due to shifting
stock_df.dropna(inplace=True)

# Display the DataFrame
stock_df




[*********************100%%**********************]  1 of 1 completed


Date        Open       Close   Adj Close  Next_Stock_Price_1  \
0   2024-01-02  373.859985  370.869995  368.854095          370.600006   
1   2024-01-03  369.010010  370.600006  368.585571          367.940002   
2   2024-01-04  370.670013  367.940002  365.940002          367.750000   
3   2024-01-05  368.970001  367.750000  365.751038          374.690002   
4   2024-01-08  369.299988  374.690002  372.653351          375.790009   
..         ...         ...         ...         ...                 ...   
167 2024-08-30  415.600006  417.140015  417.140015          409.440002   
168 2024-09-03  417.910004  409.440002  409.440002          408.899994   
169 2024-09-04  405.910004  408.899994  408.899994          408.390015   
170 2024-09-05  407.619995  408.390015  408.390015          401.700012   
171 2024-09-06  409.059998  401.700012  401.700012          405.720001   

     Next_Stock_Price_2  
0            367.940002  
1            367.750000  
2            374.690002  
3            375.790009  
4            382.769989  
..                  ...  
167          408.899994  
168          408.390015  
169          401.700012  
170          405.720001  
171          414.200012  

[172 rows x 6 columns]

## Checking for missing values

In [144]:
# Extract the date in 'yyyy-mm-dd' format
stock_df['Date'] = stock_df['Date'].dt.strftime('%d/%m/%Y')

In [145]:
# Check for missing values in the DataFrame
missing_values = stock_df.isna().sum()

# Display the count of missing values for each column
print(missing_values)


Date                  0
Open                  0
Close                 0
Adj Close             0
Next_Stock_Price_1    0
Next_Stock_Price_2    0
dtype: int64


## Ensuring the date is date time format

In [155]:
# Ensure 'Date' is in datetime format
#stock_df['Date'] = pd.to_datetime(stock_df['Date'])

# Verify data types of all columns
data_types = stock_df.dtypes

# Display the data types
print("Data types of columns:")
print(data_types)


Data types of columns:
Date                  datetime64[ns]
Open                         float64
Close                        float64
Adj Close                    float64
Next_Stock_Price_1           float64
Next_Stock_Price_2           float64
date                  datetime64[ns]
dtype: object


In [156]:
#stock_df.to_csv('MSFT_STOCK.csv',index=False)

In [157]:
# Ensure the 'date' column is of type datetime.date
stock_df['date'] = pd.to_datetime(data['date'])

In [158]:
stock_df.dtypes

Date                  datetime64[ns]
Open                         float64
Close                        float64
Adj Close                    float64
Next_Stock_Price_1           float64
Next_Stock_Price_2           float64
date                  datetime64[ns]
dtype: object

In [159]:
import pandas as pd


news_df = data

# Ensure 'Date' and 'date' columns are in datetime format
stock_df['Date'] = pd.to_datetime(stock_df['date'])
news_df['date'] = pd.to_datetime(news_df['date'])

# Merge the DataFrames on 'Date' and 'date'
merged_df = pd.merge(stock_df, news_df, left_on='Date', right_on='date', how='inner')


In [162]:
merged_df.head()

Date        Open       Close   Adj Close  Next_Stock_Price_1  \
0 2024-08-31  373.859985  370.869995  368.854095          370.600006   
1 2024-08-31  373.859985  370.869995  368.854095          370.600006   
2 2024-08-12  369.010010  370.600006  368.585571          367.940002   
3 2024-08-12  369.010010  370.600006  368.585571          367.940002   
4 2024-08-12  369.010010  370.600006  368.585571          367.940002   

   Next_Stock_Price_2     date_x               source            author  \
0          367.940002 2024-08-31                Wired  Eric Ravenscraft   
1          367.940002 2024-08-31                 CNET     Adrian Marlow   
2          367.750000 2024-08-12            The Verge        Tom Warren   
3          367.750000 2024-08-12  Yahoo Entertainment        Steve Dent   
4          367.750000 2024-08-12         Hipertextual      Luis Miranda   

                                               title  \
0  Only Microsoft Can Save Windows-Powered Handhe...   
1  You Have Less Than Two Days To Nab Microsoft's...   
2  Microsoft confirms its Paint 3D app is being d...   
3  PUBG: Battlegrounds publisher acquires Hi-Fi R...   
4  Microsoft eliminará otra aplicación clásica de...   

                                         description  \
0  The dream of running all your Steam, Epic, and...   
1  Enjoy Microsoft Visual Studio for a massive 92...   
2  Microsoft is removing its Paint 3D app from it...   
3  Under a cloud of studio closures and layoffs, ...   
4  Microsoft está por eliminar una de las version...   

                                                 url         publishedAt  \
0  https://www.wired.com/story/only-microsoft-can... 2024-08-31 13:00:00   
1  https://www.cnet.com/deals/you-have-less-than-... 2024-08-31 12:34:00   
2  https://www.theverge.com/2024/8/12/24218450/mi... 2024-08-12 16:08:27   
3  https://consent.yahoo.com/v2/collectConsent?se... 2024-08-12 12:00:05   
4  http://hipertextual.com/2024/08/microsoft-elim... 2024-08-12 17:10:47   

                                             content     date_y  
0  Microsoft also requires some software to come ... 2024-08-31  
1  Labor Day is here and that means there are dis... 2024-08-31  
2  Image: The Verge\r\n\n \n\n Paint 3D, once Mic... 2024-08-12  
3  If you click 'Accept all', we and our partners... 2024-08-12  
4  Microsoft está por eliminar una de las version... 2024-08-12

In [163]:
merged_df.dtypes

Date                  datetime64[ns]
Open                         float64
Close                        float64
Adj Close                    float64
Next_Stock_Price_1           float64
Next_Stock_Price_2           float64
date_x                datetime64[ns]
source                        object
author                        object
title                         object
description                   object
url                           object
publishedAt           datetime64[ns]
content                       object
date_y                datetime64[ns]
dtype: object

In [164]:
# Check for missing values in the merged DataFrame
missing_values = merged_df.isna().sum()

# Display the missing values for each column
print("Missing values in each column of the merged DataFrame:")
print(missing_values)


Missing values in each column of the merged DataFrame:
Date                  0
Open                  0
Close                 0
Adj Close             0
Next_Stock_Price_1    0
Next_Stock_Price_2    0
date_x                0
source                0
author                0
title                 0
description           0
url                   0
publishedAt           0
content               0
date_y                0
dtype: int64


In [165]:
#merged_df.to_csv('MSFT_stock_news.csv', index=False)

# Sentiment Analysis Model

## Sentiment Analysis Library 

* Textblob: Simple and quick for polarity analysis
* VADER: Optimize for social media and news text
* Transformers (Hugging Face): for more advance, fine tuned model


So we will use the VADER LIBRARY because they analyse news text

In [166]:
merged_df.head(1)

Date        Open       Close   Adj Close  Next_Stock_Price_1  \
0 2024-08-31  373.859985  370.869995  368.854095          370.600006   

   Next_Stock_Price_2     date_x source            author  \
0          367.940002 2024-08-31  Wired  Eric Ravenscraft   

                                               title  \
0  Only Microsoft Can Save Windows-Powered Handhe...   

                                         description  \
0  The dream of running all your Steam, Epic, and...   

                                                 url         publishedAt  \
0  https://www.wired.com/story/only-microsoft-can... 2024-08-31 13:00:00   

                                             content     date_y  
0  Microsoft also requires some software to come ... 2024-08-31

In [167]:
!pip install vaderSentiment


[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [168]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

df = merged_df

# Initialize VADER analyzer
analyzer = SentimentIntensityAnalyzer()

# Define a function to calculate sentiment
def get_vader_sentiment(text):
    score = analyzer.polarity_scores(text)
    return score['compound']

# Apply the function to the 'content' column
df['vader_sentiment'] = df['content'].apply(lambda x: get_vader_sentiment(str(x)))

# Check the results
print(df[['title', 'vader_sentiment']])

                                                 title  vader_sentiment
0    Only Microsoft Can Save Windows-Powered Handhe...           0.6705
1    You Have Less Than Two Days To Nab Microsoft's...           0.3182
2    Microsoft confirms its Paint 3D app is being d...          -0.4874
3    PUBG: Battlegrounds publisher acquires Hi-Fi R...           0.5423
4    Microsoft eliminará otra aplicación clásica de...          -0.2960
..                                                 ...              ...
449  5 Hopes And Predictions For Gamescom’s Opening...           0.4019
450  It took Microsoft 30 years to change this Wind...           0.2023
451  German Cyber Agency Wants Changes in Microsoft...           0.2732
452  Microsoft ya no te hostigará para que actualic...           0.0516
453  Microsoft backs off on pressuring Windows 10 u...          -0.4215

[454 rows x 2 columns]


In [169]:
df[['title', 'vader_sentiment']]

title  vader_sentiment
0    Only Microsoft Can Save Windows-Powered Handhe...           0.6705
1    You Have Less Than Two Days To Nab Microsoft's...           0.3182
2    Microsoft confirms its Paint 3D app is being d...          -0.4874
3    PUBG: Battlegrounds publisher acquires Hi-Fi R...           0.5423
4    Microsoft eliminará otra aplicación clásica de...          -0.2960
..                                                 ...              ...
449  5 Hopes And Predictions For Gamescom’s Opening...           0.4019
450  It took Microsoft 30 years to change this Wind...           0.2023
451  German Cyber Agency Wants Changes in Microsoft...           0.2732
452  Microsoft ya no te hostigará para que actualic...           0.0516
453  Microsoft backs off on pressuring Windows 10 u...          -0.4215

[454 rows x 2 columns]

The vader_sentiment scores you see in your results represent the overall sentiment of each news title as calculated by the VADER (Valence Aware Dictionary and sEntiment Reasoner) model. Here's what each score means:

VADER Compound Score: This is a normalized, weighted score between -1 (most extreme negative) and +1 (most extreme positive), which represents the overall sentiment of the text.

Interpreting the Results:
Oakland Police Are Towing Teslas for Crime Scene Footage:

Score: -0.7906
This is a strongly negative sentiment. The model interprets the content as conveying negative information, likely due to words like "towing" or "crime," which may be perceived negatively.

The X TV app is out now and yes, it really sucks:
Score: 0.5423
This is a positive sentiment, which may seem counterintuitive given the use of "sucks." However, VADER may misinterpret this due to sentence structure or sarcasm. Manual review or use of a more sophisticated model for detecting sarcasm might be needed here.


Tesla boosts sales in China — and may be planning a 6-seat Model Y:
Score: 0.7506
This is a strongly positive sentiment. The positive score reflects the optimistic tone of the text, highlighting sales growth and potential new plans, which are usually framed positively.

## Add Sentiment category column for positive, negative or neutral

In [170]:
sentiment_df = df

# Function to classify sentiment
def classify_sentiment(vader_score):
    if vader_score > 0:
        return 'positive'
    elif vader_score < 0:
        return 'negative'
    else:
        return 'nuetral'

# Apply the function to create a new column
sentiment_df['sentiment_category'] = sentiment_df['vader_sentiment'].apply(classify_sentiment)

In [171]:
# Remove the 'date_x' column
sentiment_df = sentiment_df.drop(columns=['date_x'])

# Rename the 'date_y' column to 'date'
sentiment_df = sentiment_df.rename(columns={'date_y': 'date'})

sentiment_df = sentiment_df.rename(columns={'Close': 'Stock_Price'})

In [173]:
sentiment_df.head()

Date        Open  Stock_Price   Adj Close  Next_Stock_Price_1  \
0 2024-08-31  373.859985   370.869995  368.854095          370.600006   
1 2024-08-31  373.859985   370.869995  368.854095          370.600006   
2 2024-08-12  369.010010   370.600006  368.585571          367.940002   
3 2024-08-12  369.010010   370.600006  368.585571          367.940002   
4 2024-08-12  369.010010   370.600006  368.585571          367.940002   

   Next_Stock_Price_2               source            author  \
0          367.940002                Wired  Eric Ravenscraft   
1          367.940002                 CNET     Adrian Marlow   
2          367.750000            The Verge        Tom Warren   
3          367.750000  Yahoo Entertainment        Steve Dent   
4          367.750000         Hipertextual      Luis Miranda   

                                               title  \
0  Only Microsoft Can Save Windows-Powered Handhe...   
1  You Have Less Than Two Days To Nab Microsoft's...   
2  Microsoft confirms its Paint 3D app is being d...   
3  PUBG: Battlegrounds publisher acquires Hi-Fi R...   
4  Microsoft eliminará otra aplicación clásica de...   

                                         description  \
0  The dream of running all your Steam, Epic, and...   
1  Enjoy Microsoft Visual Studio for a massive 92...   
2  Microsoft is removing its Paint 3D app from it...   
3  Under a cloud of studio closures and layoffs, ...   
4  Microsoft está por eliminar una de las version...   

                                                 url         publishedAt  \
0  https://www.wired.com/story/only-microsoft-can... 2024-08-31 13:00:00   
1  https://www.cnet.com/deals/you-have-less-than-... 2024-08-31 12:34:00   
2  https://www.theverge.com/2024/8/12/24218450/mi... 2024-08-12 16:08:27   
3  https://consent.yahoo.com/v2/collectConsent?se... 2024-08-12 12:00:05   
4  http://hipertextual.com/2024/08/microsoft-elim... 2024-08-12 17:10:47   

                                             content       date  \
0  Microsoft also requires some software to come ... 2024-08-31   
1  Labor Day is here and that means there are dis... 2024-08-31   
2  Image: The Verge\r\n\n \n\n Paint 3D, once Mic... 2024-08-12   
3  If you click 'Accept all', we and our partners... 2024-08-12   
4  Microsoft está por eliminar una de las version... 2024-08-12   

   vader_sentiment sentiment_category  
0           0.6705           positive  
1           0.3182           positive  
2          -0.4874           negative  
3           0.5423           positive  
4          -0.2960           negative

## Rechecking missing values

In [174]:
duplicate_rows = sentiment_df[sentiment_df.duplicated()]
print(f"Number of duplicate rows: {len(duplicate_rows)}")

# Display duplicate rows if any
if not duplicate_rows.empty:
    print("Duplicate rows:")
    print(duplicate_rows)

# Check for missing values in each column
missing_values = sentiment_df.isnull().sum()
print("\nMissing values in each column:")
print(missing_values)

Number of duplicate rows: 0

Missing values in each column:
Date                  0
Open                  0
Stock_Price           0
Adj Close             0
Next_Stock_Price_1    0
Next_Stock_Price_2    0
source                0
author                0
title                 0
description           0
url                   0
publishedAt           0
content               0
date                  0
vader_sentiment       0
sentiment_category    0
dtype: int64


In [ ]:
sentiment_df.to_csv(f'{QUERY}_stock_sentiment_analysis.csv',index=False)